### PR TITLE
Add a src/platform/Darwin and get it to link with the iOS demo app

### PR DESCRIPTION
**Pull request by vivien-apple**
#### Problem

With the current setup a lot of logic code will lives into the Darwin Framework. An other approach would be to add a new Darwin platform under src/platform and use it for the underlying logic, such as the event loop control, the bluetooth implementation, ...

Then this code could be glued to the iOS app which will then call InitChipStack at some point.

As an example this approach will contradict #1460 where the BLEPlatformDelegate code lives into the Framework. 
Here it will lives into src/platform/Darwin/BleManagerImpl.cpp 

Also, in #1401 I removed the service events code from the Darwin Framework to use the one from CHIPDeviceController. One of the thing I have not added back yet is the lock/unlock mechanism that was used in order to not keep the lock for the whole duration of the `select`. The logical followup for this, if this PR ever lands, would be to use DeviceLayer::PlatformMgr().LockChipStack(); and DeviceLayer::PlatformMgr().UnlockChipStack(); in ChipDeviceController::ServiceEvents() to add it back.

So basically the idea is to move some of the logic from src/darwin/Framework to src/platform/Darwin

### CODE REVIEW

**Comment by vivien-apple** [comment]
I we ever did that I may need some help from @pan-apple about the Mbedtls thing. In order to get the code to link I add to force 2 files to go into the build that were normally ignored if CONFIG_DEVICE_LAYER is set. I *think* we may need to some sort of special config files for Mbedtls but it is not crystal clear to me which flags to set/unset.

**Review by andy31415 — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

**Review by gerickson — APPROVED** [review]

**Comment by vivien-apple** [comment]
Pushed an updated version since there was an error initializing the configuration manager.

**Review by BroderickCarlin — APPROVED** [review]

File: src/platform/Darwin/CHIPDevicePlatformEvent.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+{
+    kPlatformDarwinEvent = kRange_InternalPlatformSpecific,
+    kPlatformDarwinBleC1WriteEvent,
+    kPlatformDarwinBleOutOfBuffersEvent,
+};
+
+} // namespace DeviceEventType
+
+/**
+ * Represents platform-specific event information for Linux platforms.
```
Comment (sagar-apple) [inline src/platform/Darwin/CHIPDevicePlatformEvent.h:61]: Darwin * platforms

File: src/platform/Darwin/ConfigurationManagerImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+bool ConfigurationManagerImpl::_CanFactoryReset()
+{
+    // TODO(#742): query the application to determine if factory reset is allowed.
+    return true;
+}
+
+void ConfigurationManagerImpl::_InitiateFactoryReset()
+{
+    PlatformMgr().ScheduleWork(DoFactoryReset);
```
Comment (sagar-apple) [inline src/platform/Darwin/ConfigurationManagerImpl.cpp:71]: Some of these APIs are only useful for an accessory. We are not planning for Darwin based accessories. A "Factory" reset has no meaning on a Controller. Maybe we'll have to go stub out some of this.

File: src/platform/Darwin/ConfigurationManagerImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+CHIP_ERROR ConfigurationManagerImpl::_WritePersistedStorageValue(::chip::Platform::PersistedStorage::Key key, uint32_t value)
+{
+    PosixConfig::Key configKey{ kConfigNamespace_ChipCounters, key };
+    return WriteConfigValue(configKey, value);
+}
+
+#if CHIP_DEVICE_CONFIG_ENABLE_WIFI_STATION
+CHIP_ERROR ConfigurationManagerImpl::GetWiFiStationSecurityType(Profiles::NetworkProvisioning::WiFiSecurityType & secType)
```
Comment (sagar-apple) [inline src/platform/Darwin/ConfigurationManagerImpl.cpp:93]: No need to support this on darwin at all right? :)

**Review by woody-apple — APPROVED** [review]
